pkg/bsql/query.go

```diff
@@ -201,6 +201,67 @@ func (s *SQLSyncer) setupPagination(ctx context.Context, pToken *pagination.Toke
 	return ret, nil
 }
 
+func (s *SQLSyncer) prepareProvisioningQuery(ctx context.Context, query string, vars map[string]any) (string, []interface{}, error) {
+	var qArgs []interface{}
+
+	var parseErr error
+	updatedQuery := queryOptRegex.ReplaceAllStringFunc(query, func(token string) string {
+		key := strings.ToLower(strings.TrimSuffix(strings.TrimPrefix(token, "?<"), ">"))
+
+		if v, ok := vars[key]; ok {
+			qArgs = append(qArgs, v)
+		} else {
+			parseErr = errors.Join(parseErr, fmt.Errorf("unknown token %s", token))
+			return token
+		}
+
+		return s.getNextPlaceholder(qArgs)
+	})
+	if parseErr != nil {
+		return "", nil, parseErr
+	}
+	return updatedQuery, qArgs, nil
+}
+
+func (s *SQLSyncer) runProvisioningQueries(ctx context.Context, queries []string, vars map[string]any) error {
+	l := ctxzap.Extract(ctx)
+
+	tx, err := s.db.Begin()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := tx.Rollback(); err != nil {
+			l.Error("failed to rollback provisioning queries", zap.Error(err))
+		}
+	}()
+
+	for _, q := range queries {
+		q, qArgs, err := s.prepareProvisioningQuery(ctx, q, vars)
+		if err != nil {
+			return err
+		}
+
+		result, err := tx.ExecContext(ctx, q, qArgs...)
+		if err != nil {
+			return err
+		}
+
+		rowsAffected, err := result.RowsAffected()
+		if err != nil {
+			l.Error("failed to get rows affected", zap.Error(err))
+		}
+
+		l.Debug("query executed", zap.String("query", q), zap.Any("args", qArgs), zap.Int64("rows_affected", rowsAffected))
```

Comment:
Did we want to have a warning message or possibly return an error and rollback if rowsAffected is greater than some threshold? Maybe... 1 by default?

Reply:
For now, I've made it so that we will fail and rollback if a query affects more than one row. This is pretty restrictive, but I think I'd prefer to discover a usecase before allowing it. This feels much safer in general. Also done in [fc3935d](https://github.com/ConductorOne/baton-sql/pull/31/commits/fc3935dbae8e51aa0e228c785b23211e7f9c97d1)